File: Source/OCLayout/UIView+JuConstraint.m
//
//  NSObject+JuConstraint.m
//  JuConsout
//
//  Created by Juvid on 16/7/19.
//  Copyright © 2016年 Juvid(zhutianwei). All rights reserved.
//

#import "UIView+JuConstraint.h"
#import <objc/runtime.h>
@implementation UIView (JuConstraint)
-(void)setJu_Constraints:(NSMutableArray *)ju_Constraints{
    objc_setAssociatedObject(self, @selector(ju_Constraints), ju_Constraints, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}
-(NSMutableArray *)ju_Constraints{
    return objc_getAssociatedObject(self, @selector(ju_Constraints));
}
-(void)setJu_Priority:(UILayoutPriority )ju_Priority{
     objc_setAssociatedObject(self, @selector(ju_Priority), @(ju_Priority), OBJC_ASSOCIATION_COPY);
}
-(UILayoutPriority)ju_Priority{
     return [objc_getAssociatedObject(self, @selector(ju_Priority)) floatValue];
}
#pragma mark 私有方法

-(void)juCompareSameCons:(NSLayoutConstraint *)currentCons{
    NSLayoutConstraint *lastCons=[self juConstrain:currentCons.juLayType withPry:currentCons.priority];
    if (lastCons) {
        if ([self.superview.constraints containsObject:lastCons]) {
            [self.superview removeConstraint:lastCons];
            [self.ju_Constraints removeObject:lastCons];
        }
        else if([self.constraints containsObject:lastCons]){
            [self removeConstraint:lastCons];
            [self.ju_Constraints removeObject:lastCons];
        }
    }
}
-(NSLayoutConstraint *)juConstrain:(JuLayoutType)juLayType withPry:(UILayoutPriority )priority{
    for (NSLayoutConstraint *lastCons in self.ju_Constraints) {
        if (lastCons.juLayType==juLayType&&lastCons.priority==priority) {
            return lastCons;
        }
    }
    return nil;
}
-(NSLayoutConstraint *)juConstrain:(JuLayoutType)juLayType{
    if (self.ju_Priority==0.00) self.ju_Priority=UILayoutPriorityRequired;
    NSLayoutConstraint *layoutCons=[self juConstrain:juLayType withPry:self.ju_Priority];
    self.ju_Priority=UILayoutPriorityRequired;
    return layoutCons;
}



-(NSLayoutConstraint *)ju_Lead{
    return [self juConstrain:JuLayoutLead];
}
-(NSLayoutConstraint *)ju_Trail{
   return  [self juConstrain:JuLayoutTrail];
}
-(NSLayoutConstraint *)ju_Top{
     return  [self juConstrain:JuLayoutTop];
}
-(NSLayoutConstraint *)ju_Bottom{
    return  [self juConstrain:JuLayoutBottom];
}
-(NSLayoutConstraint *)ju_CenterX{
   return  [self juConstrain:JuLayoutCenterX];
}
-(NSLayoutConstraint *)ju_CenterY{
     return  [self juConstrain:JuLayoutCenterY];
}
-(NSLayoutConstraint *)ju_Width{
    return  [self juConstrain:JuLayoutWidth];
}
-(NSLayoutConstraint *)ju_Height{
    return  [self juConstrain:JuLayoutHeight];
}
-(NSLayoutConstraint *)ju_AspectWH{
    return  [self juConstrain:JuLayoutAspectWH];
}

-(void)setJu_Trail:(CGFloat)value{
    NSLayoutConstraint *layout=[self juConstrain:JuLayoutTrail];
    layout.constant=-value;
}

-(void)setJu_Bottom:(CGFloat)value{
    NSLayoutConstraint *layout=[self juConstrain:JuLayoutBottom];
    layout.constant=-value;
}

-(UIView *)ju_PryHigh{
    self.ju_Priority=UILayoutPriorityDefaultHigh;
   return self;
}
-(UIView *)ju_PryLow{
    self.ju_Priority=UILayoutPriorityDefaultLow;
   return self;
}
-(UIView *)ju_PryLevel{
    self.ju_Priority=UILayoutPriorityFittingSizeLevel;
    return self;
}
-(UIView *(^)(float priority))ju_PryEqual{
    return ^UIView *(float priority){
        self.ju_Priority=priority;
       return self;
    };
}
@end

@implementation NSLayoutConstraint (JuConstraint)
-(void)setJuLayType:(JuLayoutType)juLayType{
    objc_setAssociatedObject(self, @selector(juLayType), @(juLayType), OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}
-(JuLayoutType)juLayType{
    return [objc_getAssociatedObject(self, @selector(juLayType)) integerValue];
}
@end

File: JuLayout/SwiftLayout/UIView+JuConstraint.swift
//
//  UIView+JuConstraint.swift
//  JuSLayout
//
//  Created by Juvid on 2016/9/16.
//  Copyright © 2016年 Juvid(zhutianwei). All rights reserved.
//

import UIKit
public enum JuSLayoutType : Int {
    case lead       = 0 ///< 左约束
    case trail      = 1 ///< 右约束
    case centerX    = 2///< x中间约束

    case top        = 3   ///< 顶端约束（包括baseLine）
    case bottom     = 4  ///< 底端约束
    case centerY    = 5 ///< y中间约束（包括baseLine）

    case width      = 6   ///< 宽度约束
    case height     = 7  ///< 高度约束
    case aspectWH   = 8  ///< 宽度高度比例约束
}
private var juConstraints: Void?
private var juPriority : Void?
private var ju_LayType :Void?
extension UIView{
    var ju_Constraints: NSMutableArray? {
        get {
            return objc_getAssociatedObject(self, &juConstraints) as? NSMutableArray
        }
        set(newValue) {
            objc_setAssociatedObject(self, &juConstraints, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }
    var ju_Priority : UILayoutPriority?{
        get {
            return objc_getAssociatedObject(self, &juPriority) as? UILayoutPriority
        }
        set(newValue) {
            objc_setAssociatedObject(self, &juPriority, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }

    func juCompareSameCons(_ currentCons:NSLayoutConstraint?)  {
        let layoutCons = self.juConstrain(currentCons?.juLayType, priority: currentCons?.priority)
        if layoutCons != nil {
            if (self.superview?.constraints.contains(layoutCons!)) == true {
                self.superview?.removeConstraint(layoutCons!)
                self.ju_Constraints?.remove(layoutCons!)
            }else if(self.constraints.contains(layoutCons!) == true){
                self.removeConstraint(layoutCons!)
                self.ju_Constraints?.remove(layoutCons!)
            }
        }
    }
    func juConstrain(_ juLayType:JuSLayoutType?,priority:UILayoutPriority?) -> NSLayoutConstraint? {
        if ((ju_Constraints?.count) != nil) {
            for constraint in ju_Constraints! {
                let lastCons = constraint as! NSLayoutConstraint
                if (lastCons.juLayType==juLayType&&lastCons.priority==priority) {
                    return lastCons;
                }
            }
        }

        return nil
    }
    func juConstrain(_ juLayType:JuSLayoutType?) -> NSLayoutConstraint? {
        if ju_Priority == 0.00 {
            self.ju_Priority = UILayoutPriorityRequired;
        }
        let layoutCons = self.juConstrain(juLayType, priority:ju_Priority)
        self.ju_Priority=UILayoutPriorityRequired;
        return layoutCons

    }

    var ju_Lead :NSLayoutConstraint?{
        return self.juConstrain(JuSLayoutType.lead)
    }
    var ju_Trail :NSLayoutConstraint?{
        return self.juConstrain(JuSLayoutType.trail)
    }
    var ju_Top :NSLayoutConstraint?{
        return self.juConstrain(JuSLayoutType.top)
    }
    var ju_Bottom :NSLayoutConstraint?{
        return self.juConstrain(JuSLayoutType.bottom)
    }
    var ju_CenterX :NSLayoutConstraint?{
        return self.juConstrain(JuSLayoutType.centerX)
    }
    var ju_CenterY :NSLayoutConstraint?{
        return self.juConstrain(JuSLayoutType.centerY)
    }
    var ju_Width :NSLayoutConstraint?{
        return self.juConstrain(JuSLayoutType.width)
    }
    var ju_Height :NSLayoutConstraint?{
        return self.juConstrain(JuSLayoutType.height)
    }
    var ju_AspectWH :NSLayoutConstraint?{
        return self.juConstrain(JuSLayoutType.aspectWH)
    }
    
    /// 查找优先级（基本不用）
    var ju_PriHigh : UIView{
        self.ju_Priority=UILayoutPriorityDefaultHigh;
        return self;
    }
    var ju_PriLow : UIView{
        self.ju_Priority=UILayoutPriorityDefaultLow;
        return self;
    }
    var ju_PriLevel : UIView{
        self.ju_Priority=UILayoutPriorityFittingSizeLevel;
        return self;
    }
    func ju_PriEqual(_ priority:Float) -> UIView {
        self.ju_Priority=priority;
        return self;
    }

    
    /// 查找内容优先级
    var ju_HorConHugPri :UILayoutPriority?{
        return self.contentHuggingPriority(for: UILayoutConstraintAxis.horizontal)
    }
    var ju_VerConHugPri :UILayoutPriority?{
        return self.contentHuggingPriority(for: UILayoutConstraintAxis.vertical)
    }
    var ju_VerConComResPri :UILayoutPriority?{
        return self.contentCompressionResistancePriority(for: UILayoutConstraintAxis.vertical)
    }
    var ju_HorConComResPri :UILayoutPriority?{
        return self.contentCompressionResistancePriority(for: UILayoutConstraintAxis.horizontal)
    }
//    var ju_Constraints :NSMutableArray?


}


extension NSLayoutConstraint{
    var juLayType: JuSLayoutType? {
        get {
            return objc_getAssociatedObject(self, &ju_LayType) as? JuSLayoutType
        }
        set(newValue) {
            objc_setAssociatedObject(self, &ju_LayType, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }
}
